dnf/util.py

```diff
@@ -642,33 +643,104 @@ def _is_file_pattern_present(specs):
     return False
 
 
-def _is_bootc_host():
-    """Returns true is the system is managed as an immutable container, false
-    otherwise."""
-    ostree_booted = "/run/ostree-booted"
-    return os.path.isfile(ostree_booted)
-
-
-def _is_bootc_unlocked():
-    """Check whether /usr is writeable, e.g. if we are in a normal mutable
-    system or if we are in a bootc after `bootc usr-overlay` or `ostree admin
-    unlock` was run."""
+class _BootcSystem:
     usr = "/usr"
-    return os.access(usr, os.W_OK)
-
-
-def _bootc_unlock():
-    """Set up a writeable overlay on bootc systems."""
+    CLONE_NEWNS = 0x00020000 # defined in linux/include/uapi/linux/sched.h
 
-    if _is_bootc_unlocked():
-        return
-
-    unlock_command = ["bootc", "usr-overlay"]
-
-    try:
-        completed_process = subprocess.run(unlock_command, text=True)
-        completed_process.check_returncode()
-    except FileNotFoundError:
-        raise dnf.exceptions.Error(_("bootc command not found. Is this a bootc system?"))
-    except subprocess.CalledProcessError:
-        raise dnf.exceptions.Error(_("Failed to unlock system: %s", completed_process.stderr))
+    def __init__(self):
+        if not self.is_bootc_system():
+            raise RuntimeError(_("Not running on a bootc system."))
+
+        import gi
+        self._gi = gi
+
+        gi.require_version("OSTree", "1.0")
+        from gi.repository import OSTree
+
+        self._OSTree = OSTree
+
+        self._sysroot = self._OSTree.Sysroot.new_default()
+        assert self._sysroot.load(None)
+
+        self._booted_deployment = self._sysroot.require_booted_deployment()
+        assert self._booted_deployment is not None
+
+    @staticmethod
+    def is_bootc_system():
+        """Returns true is the system is managed as an immutable container, false
+        otherwise."""
+        ostree_booted = "/run/ostree-booted"
+        return os.path.isfile(ostree_booted)
+
+    @classmethod
+    def is_writable(cls):
+        """Returns true if and only if /usr is writable."""
+        return os.access(cls.usr, os.W_OK)
+
+    def _get_unlocked_state(self):
+        return self._booted_deployment.get_unlocked()
+
+    def is_unlocked_transient(self):
+        """Returns true if and only if the bootc system is unlocked in a
+        transient state, i.e. a overlayfs is mounted as read-only on /usr.
+        Changes can be made to the overlayfs by remounting /usr as
+        read/write in a private mount namespace."""
+        return self._get_unlocked_state() == self._OSTree.DeploymentUnlockedState.TRANSIENT
+
+    @classmethod
+    def _set_up_mountns(cls):
+        # os.unshare is only available in Python >= 3.12
+        libc = ctypes.CDLL(dnf.const.LIBC_SONAME)
```

Comment:
Are you going to eliminate the LIBC_SONAME constant as discussed above?

Reply:
Done!